src/fsharp/ErrorLogger.fs

```diff
@@ -686,3 +686,103 @@ let internal languageFeatureNotSupportedInLibraryError (langVersion: LanguageVer
     let featureStr = langVersion.GetFeatureString langFeature
     let suggestedVersionStr = langVersion.GetFeatureVersionString langFeature
     error (Error(FSComp.SR.chkFeatureNotSupportedInLibrary(featureStr, suggestedVersionStr), m))
+
+/// This represents the thread-local state established as each task function runs as part of the build.
+///
+/// Use to reset error and warning handlers.
+type CompilationGlobalsScope(errorLogger: ErrorLogger, phase: BuildPhase) = 
+    let unwindEL = PushErrorLoggerPhaseUntilUnwind(fun _ -> errorLogger)
+    let unwindBP = PushThreadBuildPhaseUntilUnwind phase
+
+    member _.ErrorLogger = errorLogger
+    member _.Phase = phase
+
+    // Return the disposable object that cleans up
+    interface IDisposable with
+        member d.Dispose() =
+            unwindBP.Dispose()         
+            unwindEL.Dispose()
+
+[<Sealed>]
+type AsyncErrorLoggerBuilder() =
+
+    member _.Zero () : Async<unit> = 
+        async { 
+            ()
+        }
+
+    member _.Delay (f: unit -> Async<'T>) = f()   
+
+    member _.Return value =
+        async { 
+            return value
+        }
+
+    member _.ReturnFrom (computation:Async<_>) = computation
+
+    member _.Bind (computation: Async<'a>, binder: 'a -> Async<'b>) : Async<'b> =
+        async {
```

Comment:
On thinking about this more, I don't think this is the right way to soundly re-establish thread locals in limited async programming.

The only way you can do this soundly is if you identify all the *primitive* binds that may cause a thread jump.  As I understand it the only primitive bind that can do this is PostAndReplyAsync to the mailbox processor (or the AsyncWaitHandle on a ManualResetEvent if we switch to using that).

I think it's crucial we capture that invariant and make sure no one else ever does any other primitive async operations in these nodes.

One way to get the type system to capture this is to use a new type

```fsharp
type NodeCode<'T> = NodeCode of private Async<'T>
+builder
nodeCode { ... }
```
and the only bind to an `Async<'T>` available is one hidden inside this:
```fsharp
NodeCode.AwaitNode(t: Node<'T>) : NodeCode<'T> 
```

That would make me feel much more comfortable.  We can than choose to replace the implementation of NodeCode at any time, e.g. 

```fsharp
type NodeCode<'T> = NodeCode of private (CancellationToken -> Task<'T>)
```
or
```fsharp
type NodeCode<'T> = NodeCode of private Async2<'T>  // better stacks
```

and we get the guarantee there are no other thread-hopping async waits, now or ever.   We can also write plenty of asserts that `CompileThreadStatic.ErrorLogger` is never null when NodeCode is running.

Reply:
That is fair. Originally I did have it as a separate type. Perhaps, whatever this type is, is what should be used everywhere in the build to ensure we do not lose context.

Reply:
Yes, I think so - one type that constrains down the asynchrony to be precisely what we need it to be, no more, no less.  And which thus provably re-establishes context from the ground up.   And which we can adjust later to have a different control implementation (CT -> Task, Async2 etc).

Reply:
Nothing is passing yet, but I created a `GraphNode` with a `let node = GraphNodeBuilder()`. Internally, it's using a `Async<'T>` like what you described above. The `BackgroundCompiler` has calls that return a `GraphNode` and the public service API calls them with a `Async.AwaitGraphNode`.

Reply:
Also renamed `AsyncLazy` to `LazyGraphNode`.

`GraphNode` and `LazyGraphNode` live in a single module called `BuildGraph`.